controllers/vdoconfig_controller.go

```diff
@@ -197,7 +189,96 @@ func (r *VDOConfigReconciler) fetchVsphereCloudConfigItems(vdoctx vdocontext.VDO
+func (r *VDOConfigReconciler) fetchk8sVersions(vdoctx vdocontext.VDOContext) (string, error) {
+	discoveryClient, _ := discovery.NewDiscoveryClientForConfig(r.ClientConfig)
+	k8sServerVersion, err := discoveryClient.ServerVersion()
+	if err != nil {
+		vdoctx.Logger.Error(err, "Error occurred when k8s version")
+		return "", err
+	}
+	k8sVersion := k8sServerVersion.Major + "." + k8sServerVersion.Minor
+	return k8sVersion, nil
+}
+
+func (r *VDOConfigReconciler) fetchVsphereVersions(vdoctx vdocontext.VDOContext, req ctrl.Request, vdoConfig *vdov1alpha1.VDOConfig) (versions []string, err error) {
+	var vsphereCloudConfigsList []string
+	vsphereCloudConfigsList = vdoConfig.Spec.CloudProvider.VsphereCloudConfigs
+
+	if len(vsphereCloudConfigsList) <= 0 {
+		vsphereCloudConfigsList = []string{vdoConfig.Spec.StorageProvider.VsphereCloudConfig}
+	}
+	var vsphereCloudConfigItems []vdov1alpha1.VsphereCloudConfig
+	for _, vsphereCloudConfig := range vsphereCloudConfigsList {
+		vsphereCloudConfigItem, err := r.fetchVSphereCloudConfig(vdoctx, vsphereCloudConfig, req.Namespace)
+		vsphereCloudConfigItems = append(vsphereCloudConfigItems, *vsphereCloudConfigItem)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var vsphereVersions []string
+
+	for _, vsphereCloudConfig := range vsphereCloudConfigItems {
+		sess, err := r.getVcSession(vdoctx, &vsphereCloudConfig)
+		if err != nil {
+			return nil, err
+		}
+		vsphereVersions = append(vsphereVersions, sess.VsphereVersion)
+	}
+
+	return vsphereVersions, nil
+}
+
+func (r *VDOConfigReconciler) getVcSession(ctx context.Context, config *vdov1alpha1.VsphereCloudConfig) (*session.Session, error) {
+	var vcUser, vcUserPwd, datacenter string
+
+	if len(config.Spec.Credentials) > 0 {
+		vcCredsSecret := &v1.Secret{}
+		key := types.NamespacedName{
+			Namespace: VC_CREDS_SECRET_NS,
+			Name:      config.Spec.Credentials,
+		}
+
+		err := r.Get(ctx, key, vcCredsSecret)
+		if err != nil {
+			config.Status.Config = vdov1alpha1.VsphereConfigFailed
+			config.Status.Message = fmt.Sprintf("could not fetch vc credentials secret %s", config.Spec.Credentials)
+			return nil, errors.Wrapf(err, "could not fetch vc credentials secret %s", config.Spec.Credentials)
+		}
+
+		vcUser = string(vcCredsSecret.Data["username"])
+		vcUserPwd = string(vcCredsSecret.Data["password"])
+	}
+
+	if len(config.Spec.DataCenters) > 0 {
+		datacenter = config.Spec.DataCenters[0]
+	}
+
+	vcIp := config.Spec.VcIP
+	sess, err := session.GetOrCreate(ctx, vcIp, datacenter, vcUser, vcUserPwd, config.Spec.Thumbprint)
+	if err != nil {
+		config.Status.Config = vdov1alpha1.VsphereConfigFailed
+		config.Status.Message = fmt.Sprintf("Error establishing session with vcenter %s for user %s", vcIp, vcUser)
+		return nil, errors.Wrapf(err, "Error establishing session with vcenter %s for user %s", vcIp, vcUser)
+	}
+
```

Comment:
we should just return session here. the rest of the code in the function isnt required